pkg/util/winutil/eventlog/subscription/subscription.go

```diff
@@ -0,0 +1,455 @@
+// before our code reaches the ResetEvent(waitHandle). If this happens we will not see those
+// events until newer events are created and Windows once again calls SetEvent(waitHandle).
+func (q *pullSubscription) synchronizeNoMoreItems() error {
+	if !q.useNotifyChannel {
+		_ = windows.ResetEvent(windows.Handle(q.waitEventHandle))
+		return nil
+	}
+
+	// If notifyEventsAvailableLoop is blocking on WaitForMultipleObjects
+	// wake it up so we can sync on notifyNoMoreItems
+	// If notifyEventsAvailableLoop is blocking on notifyNoMoreItems channel then this is a no-op
+	windows.SetEvent(windows.Handle(q.waitEventHandle))
+	// If notifyEventsAvailableLoop is blocking on sending notifyEventsAvailable
+	// then wake/cancel it so it does not erroneously send notifyEventsAvailable.
+	select {
+	case <-q.notifyStop:
+		return fmt.Errorf("stop signal")
+	case <-q.notifyEventsAvailableLoopDone:
+		return fmt.Errorf("notify loop is not running")
+	case q.notifyNoMoreItems <- struct{}{}:
+		break
+	}
+	// Reset the events ready event so notifyEventsAvailableLoop will wait again in WaitForMultipleObjects,
+	// then write to notifyNoMoreItemsComplete to tell the loop that the event has been reset and it
+	// can safely continue.
+	_ = windows.ResetEvent(windows.Handle(q.waitEventHandle))
+	select {
+	case <-q.notifyStop:
+		return fmt.Errorf("stop signal")
+	case <-q.notifyEventsAvailableLoopDone:
+		return fmt.Errorf("notify loop is not running")
+	case q.notifyNoMoreItemsComplete <- struct{}{}:
+		break
+	}
+	return nil
+}
+
+// GetEvents returns the next available events in the subscription.
+func (q *pullSubscription) GetEvents() ([]*evtapi.EventRecord, error) {
+	// Do a non-blocking check to see if the "events available" event handle is set.
+	// We should only call EvtNext when this event is set.
+	dwWait, err := windows.WaitForSingleObject(windows.Handle(q.waitEventHandle), 0)
+	if err != nil {
+		return nil, fmt.Errorf("WaitForSingleObject failed: %v", err)
+	}
+	if dwWait != windows.WAIT_OBJECT_0 {
+		// event is not set, there are no events available
+		return nil, nil
+	}
+
+	// We supply INFINITE for the Timeout parameter but if we are at the end of the log file/there are no more events EvtNext will return,
+	// it will not block forever.
+	// https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-even6/cd4c258c-5a2c-4ba8-bce3-37eefaa416e7
+	eventRecordHandles, err := q.eventLogAPI.EvtNext(q.subscriptionHandle, q.evtNextStorage, uint(len(q.evtNextStorage)), windows.INFINITE)
+	if err == nil {
+		pkglog.Debugf("EvtNext returned %v handles", len(eventRecordHandles))
+		// got events
+		eventRecords := q.parseEventRecordHandles(eventRecordHandles)
+		return eventRecords, nil
+	} else if err == windows.ERROR_TIMEOUT {
+		// no more events
+		// TODO: Should we reset the handle? MS example says no
+		pkglog.Errorf("evtnext timeout")
+		return nil, fmt.Errorf("timeout")
```

Comment:
I'm not sure. The timeout is passed to the RPC server (EventLog service) and handled there.
https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-even6/cd4c258c-5a2c-4ba8-bce3-37eefaa416e7

Maybe if the RPC transport has a timeout it could be returned? If I remove the block then it would just fall down to the generic error case and still fail, so its probably fine to just do that.

Reply:
Not big deal, it is not important, just curious